bindings/src/main/java/io/pravega/storage/extendeds3/ExtendedS3Storage.java

```diff
@@ -403,39 +405,53 @@ private Void doConcat(SegmentHandle targetHandle, long offset, String sourceSegm
-                config.getPrefix() + sourceSegment,
-                config.getBucket(),
-                targetPath,
-                uploadId,
-                2).withSourceRange(Range.fromOffsetLength(0, objectSize));
-
-        copyResult = client.copyPart(copyRequest);
-        partEtags.add(new MultipartPartETag(copyResult.getPartNumber(), copyResult.getETag()));
-
-        //Close the upload
-        client.completeMultipartUpload(new CompleteMultipartUploadRequest(config.getBucket(),
-                targetPath, uploadId).withParts(partEtags));
-
-        client.deleteObject(config.getBucket(), config.getPrefix() + sourceSegment);
-        Duration elapsed = timer.getElapsed();
+        if (config.getSmallObjectThreshold() < si.getLength()) {
+
+            //Copy the first part
+            CopyPartRequest copyRequest = new CopyPartRequest(config.getBucket(),
+                    targetPath,
+                    config.getBucket(),
+                    targetPath,
+                    uploadId,
+                    1).withSourceRange(Range.fromOffsetLength(0, offset));
+            CopyPartResult copyResult = client.copyPart(copyRequest);
+
+            partEtags.add(new MultipartPartETag(copyResult.getPartNumber(), copyResult.getETag()));
+
+            //Copy the second part
+            S3ObjectMetadata metadataResult = client.getObjectMetadata(config.getBucket(),
+                    config.getPrefix() + sourceSegment);
+            long objectSize = metadataResult.getContentLength(); // in bytes
+
+            copyRequest = new CopyPartRequest(config.getBucket(),
+                    config.getPrefix() + sourceSegment,
+                    config.getBucket(),
+                    targetPath,
+                    uploadId,
+                    2).withSourceRange(Range.fromOffsetLength(0, objectSize));
+
+            copyResult = client.copyPart(copyRequest);
+            partEtags.add(new MultipartPartETag(copyResult.getPartNumber(), copyResult.getETag()));
 
+            //Close the upload
+            client.completeMultipartUpload(new CompleteMultipartUploadRequest(config.getBucket(),
+                    targetPath, uploadId).withParts(partEtags));
+
+            client.deleteObject(config.getBucket(), config.getPrefix() + sourceSegment);
+
+        } else {
+            try (InputStream reader = client.readObjectStream(config.getBucket(),
+                    config.getPrefix() + sourceSegment, Range.fromOffsetLength(0, si.getLength()))) {
+                client.putObject(this.config.getBucket(),
+                        targetPath,
+                        Range.fromOffsetLength(offset, si.getLength()),
+                        new BufferedInputStream(reader, Math.toIntExact(si.getLength())));
+            } catch (Exception e) {
+                throw Exceptions.sneakyThrow(e);
+            }
+            client.deleteObject(config.getBucket(), config.getPrefix() + sourceSegment);
```

Comment:
Only delete original object in case of success, right? move this line into try block.

Reply:
Delete operation may fail too.  Need some error handling here and of course sufficient logging.

Reply:
Yes. actually this could be a nasty bug. If concat succeeds but delete fails then we are leaving target segment in a state where it is bigger than before. But for the caller the operation has failed so it will not update size. If now operation is retried it will be done with older offset which will fail again and again. Solution is to either move delete as a separate step to upper layers or not fail the concat call when delete operation fails.

Reply:
This is not as bad as you think. If the concat succeeds but the delete fails, the storage writer will reattempt. You will need to throw a `BadOffsetException` because your precondition will not hold. The StorageWriter will them enter a reconciliation state and figure out that the source segment has actually been merged in. 

The only downside is that the old segment will be left behind, occupying space but not being used.

Reply:
Okay, it makes sense then.

Reply:
I just realized that in case of delete failures, the concat operation will be retried with same offset and we'll simply overwrite the same byte range with same contents. There is a test case that tests this idempotent. If the delete failure was intermittent then eventually it will succeed. Otherwise its likely that root cause will manifest in other operations too (like object deleted by some zombie etc. in which case read will fail. Same for credential issues. )

Reply:
The other observation is S3Client being used here doesn't declare any exceptions for its methods but actually it does throw [S3Exception](https://github.com/EMCECS/ecs-object-client-java/blob/master/src/main/java/com/emc/object/s3/S3Exception.java) if it receives a response with status code other than 2xx. 

Since S3Exception is defined as a runtime exception we need to ensure the call stack handles it correctly and doesn't lead to a serious disaster as it may not assume to handle a runtime exception at all.

Reply:
- Updated to throw BadOffsetException on failure during delete. This will trigger reconciliation.
- In normal course we continue to allow concat to be idempotent.